### ISSUE TITLE
Possibility of parsing jwt parameters in IAM policy scenario to arrays

### ISSUE THREAD

**Issue by dnoelb** [minio/minio#11303]
**Is your feature request related to a problem? Please describe.**
I have one minio bucket with several directories, which represent specific projects. I want to grant a user (which is authenticated via openID) access to some of the project directories. This is done by using a policy like this:

```
    "Version": "2012-10-17",
    "Statement": [
      {
        "Action": [
          "s3:ListBucket"
        ],
        "Effect": "Allow",
        "Resource": [
          "arn:aws:s3:::insitusync"
        ],
        "Condition": {"StringLike" : {"s3:prefix" : ["", "${jwt:groups}"]}}
      }
    ]
```
  
The jwt:groups parameter contains an json array of directory-ids, like ["1/*","2/*","3/*"].

If i write the directory ids directly into the policy like below everything works fine.

`
   "Condition": {"StringLike" : {"s3:prefix" : ["", "1/*","2/*","3/*"]}}`

But if i use the jwt:groups parameter it only works with single strings and not with an array.

**Describe the solution you'd like**
I would like to directly use the jwt:groups parameter in a policy to grant access to multiple directories of a bucket.

```
    "Version": "2012-10-17",
    "Statement": [
      {
        "Action": [
          "s3:ListBucket"
        ],
        "Effect": "Allow",
        "Resource": [
          "arn:aws:s3:::insitusync"
        ],
        "Condition": {"StringLike" : {"s3:prefix" : ["${jwt:groups}"]}}
      }
    ]
```

**Describe alternatives you've considered**
I considered retrieving a new token for each project. But this leads to a more complicated solution and quite a lot of unnecessary overhead in the frontend/backend.

**Comment by footplus** [comment]
Another scenario that comes to mind is the handling of arrays at the policy claim name for a user: a token with "groups":["a","b"] could be attached a synthetic policy of the policy "a" and the policy "b"; but there may perhaps be performance and evaluation order/depth side-effects.

**Comment by harshavardhana** [comment]
>         "Condition": {"StringLike" : {"s3:prefix" : ["", "${jwt:groups}"]}}

How about `jwt:groups` provide a comma-separated value? -we can treat such values separately if they are `comma` separated - since this a fully structured JSON type we cannot do possibly support what you are asking @dnoelb 

But there is another option that we can explore is if you can index the values then they can populated 

>     "Version": "2012-10-17",
>     "Statement": [
>       {
>         "Action": [
>           "s3:ListBucket"
>         ],
>         "Effect": "Allow",
>         "Resource": [
>           "arn:aws:s3:::insitusync"
>         ],
>         "Condition": {"StringLike" : {"s3:prefix" : ["", "${jwt:groups[0]}", "${jwt:groups[1]}"...]}}
>       }
>     ]

**Comment by footplus** [comment]
I don't know about @dnoelb's scenario, but in mine, groups is a list of AD groups which do not follow a particular order.

I just though, would there be a way to delegate this kind of logic to a user-provided webhook at the STS stage ? I.e a webhook that would receive the pre-validated user's token, and would output the applicable policy for this AK/SK/ST tuple ?

This way we could perhaps do away with complex replacement inside policy templates, and permit using whatever logic the user see fit. However this requires storing the policy document somewhere, because it could be somewhat extensive.

**Comment by harshavardhana** [comment]
> I just though, would there be a way to delegate this kind of logic to a user-provided webhook at the STS stage ? I.e a webhook that would receive the pre-validated user's token, and would output the applicable policy for this AK/SK/ST tuple ?

We cannot externalize our policy enforcements, neither we can go against AWS S3 IAM style policy and introduce our own mechanisms here since no one will know how to use them

> This way we could perhaps do away with complex replacement inside policy templates, and permit using whatever logic the user see fit. However this requires storing the policy document somewhere, because it could be somewhat extensive.

Correct externalizing policy enforcement is way too much of a design change which we cannot do at this point in time.

**Comment by footplus** [comment]
Well, i was not thinking of externalizing the policy enforcement itself, which will indeed have the same issues that the now deprecated OPA feature.

No, I was rather thinking of how flexibly binding or generating policy documents for users could be done.

In my use-case, related to data exploration, every OpenID-authenticated user has virtually a different permission set, depending on the specific datasets she have to work on. So, i currently chose to write per-user policies and use a user-unique JWT claim (email for example) as the policy name. It sadly implies that I do have to write a different policy for each user (which somewhats limits the gain of adding OpenID authentication, since I then can't have a generic policy for unknown users). 

If that policy document could be generated (in existing S3 IAM language) by an external trusted process at the time the user exchanges her JWT for AK/SK/ST tuple, it could be a solution. After all, in this STS call, a policy document can even be passed by the user itself to get a more restricted token.

So what I had in mind was:
- The user calls AssumeRoleWithWebIdentity with her JWT.
- MinIO at this stage can either directly lookup the policy attribute of the JWT and return AK/SK/ST like today - or submit the user's JWT to the policy webhook.
- The policy webhook then returns the IAM-style policy document to apply for the generated AK/SK/ST. The policy document itself could result in a new temporary policy stored at the cluster, so that it could be evaluated and enforced like today.

However, I understand very well that all this is a lot of work to get right, and that it may not be a priority :) Thanks for reading me, and sorry @dnoelb for hijacking your issue.

**Comment by fcomte** [comment]
@harshavardhana : Is `${jwt:groups[0]}` suppose to work ?

```
{
 "Version": "2012-10-17",
 "Statement": [
  {
   "Effect": "Allow",
   "Action": [
    "s3:*"
   ],
   "Resource": [
    "arn:aws:s3:::${jwt:groups[0]}",
    "arn:aws:s3:::${jwt:groups[0]}/*",
    "arn:aws:s3:::${jwt:preferred_username}",
    "arn:aws:s3:::${jwt:preferred_username}/*"
   ]
  }
 ]
}
```

This rule work only for the preferred username bucket but not for the first group of the groups list.

**Comment by harshavardhana** [comment]
> @harshavardhana : Is `${jwt:groups[0]}` suppose to work ?
> 
> ```
> {
>  "Version": "2012-10-17",
>  "Statement": [
>   {
>    "Effect": "Allow",
>    "Action": [
>     "s3:*"
>    ],
>    "Resource": [
>     "arn:aws:s3:::${jwt:groups[0]}",
>     "arn:aws:s3:::${jwt:groups[0]}/*",
>     "arn:aws:s3:::${jwt:preferred_username}",
>     "arn:aws:s3:::${jwt:preferred_username}/*"
>    ]
>   }
>  ]
> }
> ```
> 
> This rule work only for the preferred username bucket but not for the first group of the groups list.
> 

No it will not work yet..  https://github.com/minio/minio/tree/master/docs/multi-user#policy-variables

We need few other changes before we can make this work.

**Comment by antonioptorres** [comment]
> > @harshavardhana : Is `${jwt:groups[0]}` suppose to work ?
> > ```
> > {
> >  "Version": "2012-10-17",
> >  "Statement": [
> >   {
> >    "Effect": "Allow",
> >    "Action": [
> >     "s3:*"
> >    ],
> >    "Resource": [
> >     "arn:aws:s3:::${jwt:groups[0]}",
> >     "arn:aws:s3:::${jwt:groups[0]}/*",
> >     "arn:aws:s3:::${jwt:preferred_username}",
> >     "arn:aws:s3:::${jwt:preferred_username}/*"
> >    ]
> >   }
> >  ]
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This rule work only for the preferred username bucket but not for the first group of the groups list.
> 
> No it will not work yet.. https://github.com/minio/minio/tree/master/docs/multi-user#policy-variables
> 
> We need few other changes before we can make this work.


Hey there! Was there any follow up related to this solution?

**Comment by harshavardhana** [comment]
> Hey there! Was there any follow up related to this solution?

No nothing yet.

**Comment by fcomte** [comment]
My workaround at the moment is to duplicate the mapping of user in groups. 
I create the following statements  in my sts policy : 

```
   {
    "Effect": "Allow",
    "Action": [
     "s3:*"
    ],
    "Resource": [
     "arn:aws:s3:::groupe-x",
     "arn:aws:s3:::groupe-x/*"
    ],
    "Condition": {
     "StringEquals": {
      "jwt:preferred_username": [
       "user-a",
       "user-b",
       "user-c"
      ]
     }
    }
   },
```  
My STS policy is heavy : 13 ko.
Is there any problem with such a big policy ?

**Comment by harshavardhana** [comment]
You can have upto 20KiB max @fcomte since policy document.

**Comment by harshavardhana** [comment]
External AuthN and External AuthZ are already GA'ed many months ago. It addresses this issue thoroughly where in you can write your own endpoints.

Additionally `array` verification is already supported now in our policy documents so the original issue is already addressed, as an enhancement PR #16538 will support 

- ldap:groups (LDAP as IDP)
- aws:groups (MinIO as IDP)

Closing this issue as complete.
